File: 数组/前缀和/2909.java
class Solution {
    public int minimumSum(int[] nums) {
        int n = nums.length;

        int[] leftMin = new int[n];
        int minSoFar = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            if (minSoFar < nums[i]) {
                leftMin[i] = minSoFar;
            } else {
                leftMin[i] = Integer.MAX_VALUE;
            }
            minSoFar = Math.min(minSoFar, nums[i]);
        }

        int[] rightMin = new int[n];
        minSoFar = Integer.MAX_VALUE;
        for (int i=n-1;i>=0;i--) {
            if (minSoFar < nums[i]) {
                rightMin[i] = minSoFar;
            }
            else {
                rightMin[i] = Integer.MAX_VALUE;
            }
            minSoFar = Math.min(minSoFar, nums[i]);
        }

        int ans=Integer.MAX_VALUE;
        for (int j=1;j<n-1;j++) {
            if (leftMin[j] != Integer.MAX_VALUE && rightMin[j] != Integer.MAX_VALUE) {
                ans = Math.min(ans, leftMin[j] + nums[j] + rightMin[j]);
            }
        }

        return (ans!=Integer.MAX_VALUE) ? ans : -1;
    }
}

File: 力扣/数组/前缀和/2909.cpp
class Solution {
public:
    int minimumSum(vector<int>& nums) {
        int n=nums.size();

        vector<int> left(n, INT_MAX);
        vector<int> right(n, INT_MAX);
        int minSoFar = INT_MAX;
        for (int i=0;i<n;i++) {
            if (minSoFar < nums[i]) {
                left[i] = minSoFar;
            }
            minSoFar = min(nums[i], minSoFar);
        }
        minSoFar = INT_MAX;
        for (int i=n-1;i>=0;i--) {
            if (minSoFar < nums[i]) {
                right[i] = minSoFar;
            }
            minSoFar = min(nums[i], minSoFar);
        }

        int ans=INT_MAX;
        for (int j=1;j<n-1;j++) {
            if (left[j]!=INT_MAX && right[j]!=INT_MAX) {
                ans = min(ans, left[j] + nums[j] + right[j]);
            }
        }

        return (ans!=INT_MAX) ? ans : -1;
    }
};
